apps/shared/avifpng.c

```diff
@@ -17,6 +20,119 @@ typedef png_bytep png_iccp_datap;
 typedef png_charp png_iccp_datap;
 #endif
 
+// Converts a hexadecimal string which contains 2-byte character representations of hexadecimal values to raw data.
+// hexString may contain values consisting of [A-F][a-f][0-9] in pairs, e.g., 7af2..., separated by any number of newlines.
+// On success the bytes are filled and AVIF_TRUE is returned.
+// AVIF_FALSE is returned if fewer than expectedLength hexadecimal pairs are converted.
+static avifBool avifHexStringToBytes(const char * hexString, size_t expectedLength, avifRWData * bytes)
+{
+    avifRWDataRealloc(bytes, expectedLength);
+    const char * src = hexString;
+    size_t actualLength = 0;
+    for (; (actualLength < expectedLength) && (*src != '\0'); ++actualLength) {
+        while (*src == '\n') {
+            ++src;
+        }
+        if (!isxdigit(src[0]) || !isxdigit(src[1])) {
+            avifRWDataFree(bytes);
+            fprintf(stderr, "Exif extraction failed: invalid token at " AVIF_FMT_ZU "\n", actualLength);
+            return AVIF_FALSE;
+        }
+        const char twoHexDigits[] = { src[0], src[1], '\0' };
+        bytes->data[actualLength] = (uint8_t)strtol(twoHexDigits, NULL, 16);
+        src += 2;
+    }
+
+    if (actualLength != expectedLength) {
+        avifRWDataFree(bytes);
+        fprintf(stderr, "Exif extraction failed: expected " AVIF_FMT_ZU " tokens but got " AVIF_FMT_ZU "\n", expectedLength, actualLength);
+        return AVIF_FALSE;
+    }
+    return AVIF_TRUE;
+}
+
+// Parses the raw profile string of profileLen characters and extracts the payload.
+static avifBool avifCopyRawProfile(const char * profile, size_t profileLength, avifRWData * payload)
+{
+    // ImageMagick formats 'raw profiles' as "\n<name>\n<length>(%8lu)\n<hex payload>\n".
```

Comment:
I think ImageMagick also prepends the `Exif\0\0` 6 byte sequence that needs to be discarded after decoding. See again the note for the eXIf chunk specification that does not include this header.

Reply:
The best I can offer is the Exif spec itself - the Exif00 header is described as required for the JPEG APP1 payload, but not for the TIFF one. Other image formats were not discussed at the time. Why ImageMagick devs decided to keep it for their custom raw PNG chunk that became so widespread is unknown (perhaps it was just more convenient to convert to/from JPEG). GIMP parser e.g. links to IM code for reference.

The PNG 1.5 extension spec linked earlier also specifically mentions the omission of this header for the more modern, standard eXIf chunk.

Reply:
I guess this Exif00 header should not be imported into the `avifImage::exif` field then. We just have to be careful when generating Jpeg and PNG images from it, if we implement that one day.

Among the AVIF images with Exif metadata that I crawled, most start with Exif00 but not all.

As a reference, [`heif_enc.cc`](https://github.com/strukturag/libheif/blob/master/examples/heif_enc.cc) does not seem to import any kind of metadata besides the color profile. The HEIF specification ISO-23008 says in section A.2.1 that the Exif blob may or may not start with the TIFF header, leaving the choice to the user of including the Exif00 tag or not.

Reply:
Yes, I think it should be removed too.

Don't know what libheif implemented, but I have recently discussed this over at libjxl as well - they also decided to follow the official HEIF and JPEG XS spec where the Exif payload always starts with just II/MM directly (with the possible 4-byte offset at the start of the box).